charts/redpanda/secrets.go

```diff
@@ -0,0 +1,731 @@
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-fs-validator", Fullname(dot)[:49]),
+			Namespace: dot.Release.Namespace,
+			Labels:    FullLabels(dot),
+		},
+		Type:       corev1.SecretTypeOpaque,
+		StringData: map[string]string{},
+	}
+
+	secret.StringData["fsValidator.sh"] = `set -e
+EXPECTED_FS_TYPE=$1
+
+DATA_DIR="/var/lib/redpanda/data"
+TEST_FILE="testfile"
+
+echo "checking data directory exist..."
+if [ ! -d "${DATA_DIR}" ]; then
+  echo "data directory does not exists, exiting"
+  exit 1
+fi
+
+echo "checking filesystem type..."
+FS_TYPE=$(df -T $DATA_DIR  | tail -n +2 | awk '{print $2}')
+
+if [ "${FS_TYPE}" != "${EXPECTED_FS_TYPE}" ]; then
+  echo "file system found to be ${FS_TYPE} when expected ${EXPECTED_FS_TYPE}"
+  exit 1
+fi
+
+echo "checking if able to create a test file..."
+
+touch ${DATA_DIR}/${TEST_FILE}
+result=$(touch ${DATA_DIR}/${TEST_FILE} 2> /dev/null; echo $?)
+if [ "${result}" != "0" ]; then
+  echo "could not write testfile, may not have write permission"
+  exit 1
+fi
+
+echo "checking if able to delete a test file..."
+
+result=$(rm ${DATA_DIR}/${TEST_FILE} 2> /dev/null; echo $?)
+if [ "${result}" != "0" ]; then
+  echo "could not delete testfile"
+  exit 1
+fi
+
+echo "passed"`
+	return secret
+}
+
+func SecretConfigurator(dot *helmette.Dot) *corev1.Secret {
+	values := helmette.Unwrap[Values](dot.Values)
+
+	internalAdvertiseAddress := fmt.Sprintf("%s.%s", "${SERVICE_NAME}", InternalDomain(dot))
+	externalAdvertiseAddress := "${SERVICE_NAME}"
+	externalDomainTemplate := ptr.Deref(values.External.Domain, "")
+	if externalDomainTemplate != "" {
+		externalAdvertiseAddress = fmt.Sprintf("%s.%s", "${SERVICE_NAME}", helmette.Tpl(externalDomainTemplate, dot))
+	}
+
+	secret := &corev1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: "v1",
+			Kind:       "Secret",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("%s-configurator", Fullname(dot)[:51]),
+			Namespace: dot.Release.Namespace,
+			Labels:    FullLabels(dot),
+		},
+		Type:       corev1.SecretTypeOpaque,
+		StringData: map[string]string{},
+	}
+	configuratorSh := []string{}
+	configuratorSh = append(configuratorSh,
+		`set -xe`,
+		`SERVICE_NAME=$1`,
+		`KUBERNETES_NODE_NAME=$2`,
+		`POD_ORDINAL=${SERVICE_NAME##*-}`,
+		"BROKER_INDEX=`expr $POD_ORDINAL + 1`", // POSIX sh-safe
+		``,
+		`CONFIG=/etc/redpanda/redpanda.yaml`,
+		``,
+		`# Setup config files`,
+		`cp /tmp/base-config/redpanda.yaml "${CONFIG}"`,
+		`cp /tmp/base-config/bootstrap.yaml /etc/redpanda/.bootstrap.yaml`,
+	)
+	if !RedpandaAtLeast_22_3_0(dot) {
+		configuratorSh = append(configuratorSh,
+			``,
+			`# Configure bootstrap`,
+			`## Not used for Redpanda v22.3.0+`,
+			`rpk --config "${CONFIG}" redpanda config set redpanda.node_id "${POD_ORDINAL}"`,
+			`if [ "${POD_ORDINAL}" = "0" ]; then`,
+			`	rpk --config "${CONFIG}" redpanda config set redpanda.seed_servers '[]' --format yaml`,
+			`fi`,
+		)
+	}
+
+	{ // Handle kafka listener
```

Comment:
😬 So the scoping guarantees of blocks do not extend to gotohelm. It might be best to factor these out into different functions just to be safe?

I see what you mean about the pain of dealing with all the slightly differently typed listeners now.

Reply:
Yeah, I wanted to get this working first - the blocks are there as a visual guide as much as anything else.

The bulk of the noise in here comes from the support for templating in advertised names, which I think we need to keep - there's a contract there that we need to confirm is working.

Reply:
Given that these are functions rather than methods, I think we might be able to refactor this to write that function once. I'll put up something to address the nits then look at that.